bigframes/session/__init__.py

```diff
@@ -690,16 +686,67 @@ def _read_gbq_table(
                 f"Current session is in {self._location} but dataset '{table.project}.{table.dataset_id}' is located in {table.location}"
             )
 
+        # ----------------------------------------
+        # Convert index_col into a list of columns
+        # ----------------------------------------
+
+        # This requires the table metadata because we might use the primary
+        # keys when constructing the index.
+        index_cols = bf_read_gbq_table.get_index_cols(
+            table=table,
+            index_col=index_col,
+        )
+
+        # -----------------------------
+        # Optionally, execute the query
+        # -----------------------------
+
+        # max_results introduces non-determinism and limits the cost on
+        # clustered tables, so fallback to a query. We do this here so that
+        # the index is consistent with tables that have primary keys, even
+        # when max_results is set.
+        # TODO(b/338419730): We don't need to fallback to a query for wildcard
+        # tables if we allow some non-determinism when time travel isn't supported.
+        if max_results is not None or bf_io_bigquery.is_table_with_wildcard_suffix(
+            query
+        ):
+            # TODO(b/338111344): If we are running a query anyway, we might as
+            # well generate ROW_NUMBER() at the same time.
+            query = bf_io_bigquery.to_query(
+                query,
+                index_cols=index_cols,
+                columns=columns,
+                filters=filters,
+                max_results=max_results,
+                # We're executing the query, so we don't need time travel for
+                # determinism.
+                time_travel_timestamp=None,
+            )
+
+            return self._read_gbq_query(
+                query,
+                index_col=index_cols,
+                columns=columns,
+                api_name="read_gbq_table",
+                use_cache=use_cache,
+            )
+
         # -----------------------------------------
         # Create Ibis table expression and validate
         # -----------------------------------------
 
         # Use a time travel to make sure the DataFrame is deterministic, even
         # if the underlying table changes.
+        # TODO(b/340540991): If a dry run query fails with time travel but
+        # succeeds without it, omit the time travel clause and raise a warning
+        # about potential non-determinism if the underlying tables are modified.
         table_expression = bf_read_gbq_table.get_ibis_time_travel_table(
-            self.ibis_client,
-            table_ref,
-            time_travel_timestamp,
+            ibis_client=self.ibis_client,
+            table_ref=table_ref,
+            index_cols=index_cols,
+            columns=columns,
+            filters=filters,
+            time_travel_timestamp=time_travel_timestamp,
         )
 
         for key in columns:
```

Comment:
@TrevorBergeron Which "validation for unknown column names'" are we missing? I still see this here.

Reply:
Would you like me to move it to before we create the ibis expression? It's possible ibis will fail with a dry run error when we try to access the columns in it now.

Reply:
I see `tests/system/small/test_session.py::test_read_gbq_w_columns[unknown_col]` failure now. Likely this is the issue. I'll move this validation to after we tech the table metadata.

Reply:
Fixed in 87365db9d494cdc6cfd18ba930dea2b788c591de